src/Storages/StorageS3Distributed.h

```diff
@@ -0,0 +1,157 @@
+using Task = std::string;
+using Tasks = std::vector<Task>;
+using TasksIterator = Tasks::iterator;
+
+class S3NextTaskResolver
+{
+public:
+    S3NextTaskResolver(QueryId query_id, Tasks && all_tasks)
+        : id(query_id)
+        , tasks(all_tasks)
+        , current(tasks.begin())
+    {}
+
+    std::string next()
+    {
+        auto it = current;
+        ++current;
+        return it == tasks.end() ? "" : *it;
+    }
+
+    std::string getId()
+    {
+        return id;
+    }
+
+private:
+    QueryId id;
+    Tasks tasks;
+    TasksIterator current;
+};
+
+using S3NextTaskResolverPtr = std::shared_ptr<S3NextTaskResolver>;
+
+class TaskSupervisor
+{
+public:
+    using QueryId = std::string;
+
+    TaskSupervisor() = default;
+
+    static TaskSupervisor & instance()
+    {
+        static TaskSupervisor task_manager;
+        return task_manager;
+    }
+
+    void registerNextTaskResolver(S3NextTaskResolverPtr resolver)
+    {
+        std::lock_guard lock(mutex);
+        auto & target = dict[resolver->getId()];
+        if (target)
+            throw Exception(fmt::format("NextTaskResolver with name {} is already registered for query {}",
+                target->getId(), resolver->getId()), ErrorCodes::LOGICAL_ERROR);
+        target = std::move(resolver);
+    }
+
+
+    Task getNextTaskForId(const QueryId & id)
+    {
+        std::lock_guard lock(mutex);
+        auto it = dict.find(id);
+        if (it == dict.end())
+            return "";
+        auto answer = it->second->next();
+        if (answer.empty())
+            dict.erase(it); 
+        return answer;
+    }
+
+private:
+    using ResolverDict = std::unordered_map<QueryId, S3NextTaskResolverPtr>;
+    ResolverDict dict;
+    std::mutex mutex;
+};
+
+class StorageS3Distributed : public ext::shared_ptr_helper<StorageS3Distributed>, public IStorage 
+{
+    friend struct ext::shared_ptr_helper<StorageS3Distributed>;
+public:
+    std::string getName() const override { return "S3Distributed"; }
+
+    Pipe read(
+        const Names & /*column_names*/,
+        const StorageMetadataPtr & /*metadata_snapshot*/,
+        SelectQueryInfo & /*query_info*/,
+        const Context & /*context*/,
+        QueryProcessingStage::Enum /*processed_stage*/,
+        size_t /*max_block_size*/,
+        unsigned /*num_streams*/) override;
+
+
+protected:
+    StorageS3Distributed(
+        IAST::Hash tree_hash_,
+        const String & address_hash_or_filename_,
+        const String & access_key_id_,
+        const String & secret_access_key_,
+        const StorageID & table_id_,
+        String cluster_name_,
+        const String & format_name_,
+        UInt64 max_connections_,
+        const ColumnsDescription & columns_,
+        const ConstraintsDescription & constraints_,
+        const Context & context_,
+        const String & compression_method_);
+
+private:
+    /// Connections from initiator to other nodes
+    std::vector<std::shared_ptr<Connection>> connections;
```

Comment:
Hedged connections and async read from socket won't be supported ...

Reply:
Deleted this code.